nerdfonts-patcher: init at v2.1.0

NomisIV opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The official nerdfonts package for NixOS doesn't support patching any font, so I made this package to do that.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

NomisIV committed:
nerdfonts-patcher: init at v2.1.0

NomisIV commented:
I'm having a problem where the two assertions fail, even though I override them when calling the package. I currently don't know how to solve this.

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,52 @@
+  buildPhase = ''
+    runHook preBuild
+    mkdir -p $out/share/fonts
+    find ${toString patchFont}/share/fonts/**/*.ttf -exec python font-patcher ${glyphFlags} -o $out/share/fonts/ {} \;
```
The modern format otf is also supported.

NomisIV replied:
Good one. I'll fix this tomorrow

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,52 @@
+  pname = fontName + "-nerd-font";
+  version = "v2.1.0";
+  
+  src = fetchFromGitHub {
```
The tarball has a size of 2.2GiB and I think only the file font-patcher needed. Could you try to pick only that file.

NomisIV replied:
I have tried this before, and it doesn't work. I think it comes down to the icon sets needing to be downloaded as well.

ck3d replied:
There is already a related issue, could you please add a link to https://github.com/ryanoasis/nerd-fonts/issues/484

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,52 @@
+    fontforge
+  ];
+
+  buildPhase = ''
```
I would be grate to package also the program font-patcher itself.

NomisIV replied:
That's a great idea, since the name of the package would suggest that, although ad-hoc patching would go against nixos declarative approach.
I could maybe focus on packaging the patcher, (and name the output store path accordingly), and then somehow integrate font patching into it?

ck3d replied:
Yes, 1. package the patcher, 2. provide a generator or wrapper derivation which can patch arbitrary font packages.

NomisIV commented:
I have now added support for patching otf as well, and I've split the derivation in two: nerdfonts-patcher-unwrapped for the patcher itself, and nerdfonts-patcher which is a sort of wrapper derivation for patching fonts. While trying to build the nerdfonts-patcher derivation I get an issue I don't know how to resolve:
```
unpacking sources
variable $src or $srcs should point to the source
```

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+, lib
+}:
+
+stdenv.mkDerivation rec {
```
Can you try to use buildPythonApplication instead of mkDerivation as described in https://nixos.org/manual/nixpkgs/stable/#python 15.19.2.2.2. buildPythonApplication function.

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+    rev = version;
+    sha256 = "sha256-j0cLZHOLyoItdvLOQVQVurY3ARtndv0FZTynaYJPR9E=";
+  };
+
```
You have to specify all dependencies of this package in buildInputs. In your case fontforge ... python itself is covered already in buildPythonApplication.

NomisIV replied:
I noticed that youtube-dl uses buildPythonPackage. Is there a difference between them?

ck3d replied:
The difference is small and should behave the same, you can find the definition here:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/python-packages.nix#L46

ck3d commented on pkgs/data/fonts/nerdfonts-patcher/wrapper.nix:
```diff
@@ -0,0 +1,48 @@
+# assert fontPackage != null;
+# assert fontName != null;
+
+stdenv.mkDerivation rec {
```
Use runCommand instead of mkDerivation.

ck3d commented:
> I have now added support for patching otf as well, and I've split the derivation in two: nerdfonts-patcher-unwrapped for the patcher itself, and nerdfonts-patcher which is a sort of wrapper derivation for patching fonts. While trying to build the nerdfonts-patcher derivation I get an issue I don't know how to resolve:
> 
> ```
> unpacking sources
> variable $src or $srcs should point to the source
> ```

mkDerivation need a src or srcs to be set. I proposed to switch to runCommand, here you do not need to set src.

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  # Only install what we actually need (font-patcher and src/glyphs/*)
+  installPhase = ''
+    mkdir -p $out
+    cp font-patcher $out
+    mkdir -p $out/src/glyphs
```
```suggestion
    mkdir -p $out/src/glyphs

    cp font-patcher $out
```

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+    '';
+    homepage = "https://nerdfonts.com/";
+    license = licenses.mit;
+    maintainers = with maintainers; [
+      nomisiv
+    ];
```
```suggestion
    maintainers = with maintainers; [ nomisiv ];
```

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/wrapper.nix:
```diff
@@ -0,0 +1,48 @@
+
+stdenv.mkDerivation rec {
+  pname = toString fontName + "-nerd-font";
+  version = "v2.1.0";
```
```suggestion
  version = "2.1.0";
```

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+stdenv.mkDerivation rec {
+  pname = "nerdfonts-patcher-unwrapped";
+  version = "v2.1.0";
```
```suggestion
  version = "2.1.0";
```

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/default.nix:
```diff
@@ -0,0 +1,42 @@
+  src = fetchFromGitHub {
+    owner = "ryanoasis";
+    repo = "nerd-fonts";
+    rev = version;
```
```suggestion
    rev = "v${version}";
```

SuperSandro2000 commented on pkgs/data/fonts/nerdfonts-patcher/wrapper.nix:
```diff
@@ -0,0 +1,48 @@
+    '';
+    homepage = "https://nerdfonts.com/";
+    license = licenses.mit;
+    maintainers = with maintainers; [
+      nomisiv
+    ];
```
```suggestion
    maintainers = with maintainers; [ nomisiv ];
```
